src/psyclone/dynamo0p3_builtins.py

```diff
@@ -212,28 +219,38 @@ def _validate(self):
         field_count = 0  # We must have one or more fields as arguments
         spaces = set()   # All field arguments must be on the same space
         for arg in self.arg_descriptors:
-            # Built-ins update fields DoF by DoF and therefore can have
-            # WRITE/READWRITE access
-            if arg.access in [AccessType.WRITE, AccessType.SUM,
-                              AccessType.READWRITE]:
-                write_count += 1
-            if arg.argument_type in LFRicArgDescriptor.VALID_FIELD_NAMES:
-                field_count += 1
-                spaces.add(arg.function_space)
+            # Check valid argument types
             if arg.argument_type not in VALID_BUILTIN_ARG_TYPES:
                 raise ParseError(
                     "In the LFRic API an argument to a built-in kernel "
                     "must be one of {0} but kernel '{1}' has an argument of "
                     "type '{2}'.".format(VALID_BUILTIN_ARG_TYPES, self.name,
                                          arg.argument_type))
+            # Check valid data types (InternalError as the relevant ParseError
+            # is caught in LFRicArgDescriptor)
             if (arg.argument_type in LFRicArgDescriptor.VALID_FIELD_NAMES and
                     arg.data_type not in VALID_BUILTIN_FIELD_DATA_TYPES):
-                raise ParseError(
-                    "In the LFRic API a field argument to a built-in kernel "
-                    "must have one of {0} as data type but kernel '{1}' has "
-                    "a field argument with data type '{2}'.".
-                    format(VALID_BUILTIN_FIELD_DATA_TYPES, self.name,
-                           arg.data_type))
+                raise InternalError(
+                    "Found an unsupported data type '{0}' for a field "
+                    "argument in kernel '{1}'. Supported data types are {2}.".
+                    format(arg.data_type, self.name,
+                           VALID_BUILTIN_FIELD_DATA_TYPES))
+            if (arg.argument_type in LFRicArgDescriptor.VALID_SCALAR_NAMES and
```

Comment:
Again, this looks like something that should be done in LFRicArgDescriptor?

Reply:
Again, I followed the logic of adding a PSyclone `InternalError` check after the relevant `ParseError` check in the `LFRicArgDescriptor`. It would be good to agree where we need them.

Reply:
I always think about "entry points". If a routine acts as an entry point and is acting on information supplied externally (such as kernel metadata) then it must do checking. If it is an 'internal' routine then it is more Pythonic to do 'duck' typing and only check things when necessary. I don't think we have this formalised anywhere though and @rupertford, @sergisiso and @hiker might have different opinions!

Reply:
As above, these checks are already done in LFRic scalar and field classes so I removed this check here.